homeassistant/components/arcam_fmj/media_player.py

```diff
@@ -0,0 +1,358 @@
+    DecodeMode2CH,
+    DecodeModeMCH,
+    IncomingAudioFormat,
+    SourceCodes,
+)
+from arcam.fmj.state import State
+
+from homeassistant import config_entries
+from homeassistant.core import callback
+from homeassistant.components.media_player import MediaPlayerDevice
+from homeassistant.components.media_player.const import (
+    MEDIA_TYPE_MUSIC,
+    SUPPORT_SELECT_SOUND_MODE,
+    SUPPORT_SELECT_SOURCE,
+    SUPPORT_TURN_ON,
+    SUPPORT_TURN_OFF,
+    SUPPORT_VOLUME_MUTE,
+    SUPPORT_VOLUME_SET,
+    SUPPORT_VOLUME_STEP,
+)
+from homeassistant.const import (
+    CONF_NAME,
+    CONF_ZONE,
+    CONF_HOST,
+    CONF_PORT,
+    SERVICE_TURN_ON,
+    STATE_OFF,
+    STATE_ON,
+)
+from homeassistant.helpers.typing import HomeAssistantType, ConfigType
+from homeassistant.helpers.service import async_call_from_config
+
+from .const import (
+    SIGNAL_CLIENT_DATA,
+    SIGNAL_CLIENT_STARTED,
+    SIGNAL_CLIENT_STOPPED,
+    DEFAULT_NAME,
+    DOMAIN_DATA_ENTRIES,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistantType,
+    config_entry: config_entries.ConfigEntry,
+    async_add_entities,
+):
+    """Set up the configuration entry."""
+    data = hass.data[DOMAIN_DATA_ENTRIES][config_entry.entry_id]
+    client = data["client"]
+    config = data["config"]
+
+    host = config[CONF_HOST]
+    port = config[CONF_PORT]
+
+    async_add_entities(
+        [
+            ArcamFmj(
+                State(client, zone),
+                zone_config.get(
+                    CONF_NAME,
+                    "{} ({}:{}) - {}".format(DEFAULT_NAME, host, port, zone),
+                ),
+                zone_config.get(SERVICE_TURN_ON),
+            )
+            for zone, zone_config in config[CONF_ZONE].items()
+        ]
+    )
+
+    return True
+
+
+class ArcamFmj(MediaPlayerDevice):
+    """Representation of a media device."""
+
+    def __init__(self, state: State, name: str, turn_on: Optional[ConfigType]):
+        """Initialize device."""
+        self._state = state
+        self._name = name
+        self._turn_on = turn_on
+        self._support = (
+            SUPPORT_SELECT_SOURCE
+            | SUPPORT_VOLUME_SET
+            | SUPPORT_VOLUME_MUTE
+            | SUPPORT_VOLUME_STEP
+            | SUPPORT_TURN_OFF
+        )
+        if state.zn == 1:
+            self._support |= SUPPORT_SELECT_SOUND_MODE
+
+    def _get_2ch(self):
+        """Return if source is 2 channel or not."""
+        audio_format, _ = self._state.get_incoming_audio_format()
+        return bool(
+            audio_format
+            in (
+                IncomingAudioFormat.PCM,
+                IncomingAudioFormat.ANALOGUE_DIRECT,
+                None,
+            )
+        )
+
+    @property
+    def unique_id(self) -> Optional[str]:
+        """Return a unique ID."""
+        return "{}_{}_{}".format(
+            self._state.client.host, self._state.client.port, self._state.zn
```

Comment:
If host is an ip addres, that doesn't fulfill the requirements for unique_id. The unique_id must be stable. IP addresses can change.

https://developers.home-assistant.io/docs/en/entity_registry_index.html#unique-id-requirements

Reply:
Indeed, but none is available. And since for annoying reasons unique ID is required for device registry, this is as good as it gets. And no MAC is not an option, since my test system is on different subnet over VPN.
Since unique ID is per domain, it's not going to conflict anything.

Reply:
If it's not stable, we shouldn't implement unique_id.

Reply:
Martin is right. The reason we have strict rules around unique ID is that if unique ID changes, all entities will be loaded with different entity IDs, and the old IDs can no longer be reused until manually removed. 

I like the fallback to mac addresses as it solves it for 99% of the users that are on the same subnet. However, I prefer real identifiers from the API .

Reply:
@balloob i do understand why strict rules on unique_id, the problem is many things have ended up needing unique_id instead of entity_id, so without it all the nice new gui features are lost.

MAC using arp lookup also breaks on a normal docker setup, as far as i know hassio runs docker (maybe it's using a bridged network instead of NAT so it probably work there).

I can remove all device_registry, config registry stuff and move back to config based friendly_name, and entity naming instead. Seems like going back to stoneage feature wise.

Reply:
Just so that there is no misunderstanding: config entry (config flow) doesn't require unique_id.

Reply:
Right. it will just not show any devices/entities linked to the integration.

Reply:
I've removed unique_id from code. It will still show an integration, just no devices or entities linked to it.